src/Data/Text/Lazy.hs

```diff
@@ -424,7 +424,23 @@ unpack ::
 #endif
   Text -> String
 unpack t = S.unstreamList (stream t)
-{-# INLINE [1] unpack #-}
+{-# NOINLINE unpack #-}
+
+foldrFB :: (Char -> b -> b) -> b -> Text -> b
+foldrFB = foldr
+{-# INLINE [0] foldrFB #-}
+
+-- List fusion rules for `unpack`:
+-- * `unpack` rewrites to `build` up till (but not including) phase 1. `build`
+--   fuses if `foldr` is applied to it.
+-- * If it doesn't fuse: In phase 1, `build` inlines to give us `foldrFB (:) []`
+--   and we rewrite that back to `unpack`.
+-- * If it fuses: In phase 0, `foldrFB` inlines and `foldr` inlines. GHC
+--   optimizes the fused code.
+{-# RULES
+"Text.Lazy.unpack"     [~1] forall t. unpack t = Exts.build (\c n -> foldrFB c n t)
```

Comment:
```suggestion
"Text.Lazy.unpack"     [~1] forall t. unpack t = Exts.build (\cons nil -> foldrFB cons nil t)
```
and same for the strict `Text`. Fusion is cryptic enough, let's give an extra hint to readers.

Reply:
Done. I went with `lcons` and `lnil` because GHC warns about `cons` shadowing the top-level `cons`.